odoo/tools/xml_utils.py

```diff
@@ -83,3 +83,74 @@ def create_xml_node(parent_node, node_name, node_value=None):
     :returns (etree._Element):
     """
     return create_xml_node_chain(parent_node, [node_name], node_value)[0]
+
+
+def cleanup_xml_node(xml_node_or_string, remove_comments=True, remove_blank_text=True, remove_blank_nodes=True, indent_level=0, indent_space="  "):
+    """Clean up the sub-tree of the provided XML node.
+
+    If the provided XML node is of type:
+    - etree._Element, it is modified in-place.
+    - string/bytes, it is first parsed into an etree._Element
+    :param xml_node_or_string (etree._Element, str): XML node (or its string/bytes representation)
+    :param remove_comments (bool): if True, removes comment nodes
+    :param remove_blank_text (bool): if True, removes whitespace-only text from nodes
+    :param remove_blank_nodes (bool): if True, removes leaf nodes with no text (iterative, depth-first, done after remove_blank_text)
+    :param indent_level (int): depth or level of node within root tree (use -1 to leave indentation as-is)
+    :param indent_space (str): string to use for indentation (use '' to remove all indentation)
+    :returns (etree._Element): clean node, same instance that was received (if applicable)
+    """
+    xml_node = xml_node_or_string
+
+    # Input is str/bytes: parser takes care of removing blanks/comments
+    if isinstance(xml_node, (str, bytes)):
+        parser = etree.XMLParser(remove_blank_text=remove_blank_text, remove_comments=remove_comments)
+        if isinstance(xml_node, str):
+            xml_node = xml_node.encode()  # misnomer: fromstring actually reads bytes
+        xml_node = etree.fromstring(xml_node, parser=parser)
+        remove_comments = False  # indicates job is done
+        # https://lxml.de/FAQ.html: remove_blank_text only removes some blank texts
+
+    # Process leaf nodes iteratively
+    # Depth-first, so any inner node may become a leaf too (if children are removed)
+    def leaf_iter(parent_node, node):
+        for child_node in node:
+            leaf_iter(node, child_node)
+
+        # Base condition: node is not root nor inner node
+        if parent_node is not None and not len(node):
+            if remove_blank_text and node.text is not None and not node.text.strip():
+                # node.text is None iff node has self-closing tag (text='' would create closing tag)
+                node.text = ''
+            if remove_comments and (node.tag is etree.Comment):
+                parent_node.remove(node)
+            elif remove_blank_nodes and not (node.text or ''):
+                parent_node.remove(node)
+
+    leaf_iter(None, xml_node)
+
+    # Fix indentation
+    if indent_level >= 0:
+        etree.indent(xml_node, level=indent_level, space=indent_space)
```

Comment:
Seems like you are not using the same version locally as the one used by runbot.
Check that `etree.indent` or whatever replacement is available in the version required (and newer)

Reply:
https://github.com/odoo/odoo/blob/15.0/requirements.txt#L17
Any idea why version 4.3.2 is used by the runbot ?
etree.indent() only exists since version 4.5, I'll look for an alternative I guess

Reply:
> https://github.com/odoo/odoo/blob/15.0/requirements.txt#L17 Any idea why version 4.3.2 is used by the runbot ?

Like all deps, it's because 4.3.2 is the current version in Debian stable: https://packages.debian.org/buster/python-lxml
Debian/Ubuntu LTS versions are our benchmark in order to keep our software installable for the whole lifecycle of the series. It is more important to make sure our users will still be able to install *maintained* versions of all our dependencies in 3 years than to have the latest versions. It's not as if XML is a new technology anyway :P